Source/WebCore/style/ElementRuleCollector.h

```diff
@@ -103,11 +103,12 @@ class ElementRuleCollector {
     void collectMatchingAuthorRules();
     void addElementInlineStyleProperties(bool includeSMILProperties);
 
-    void matchAuthorShadowPseudoElementRules();
-    void matchHostPseudoClassRules();
-    void matchSlottedPseudoElementRules();
-    void matchPartPseudoElementRules();
-    void matchPartPseudoElementRulesForScope(const Element& partMatchingElement);
+    using GetRulesFunction = Function<RuleSet*(const ScopeRuleSets&)>;
```

Comment:
sadly we can't just pass in a `RuleSet` as each of the functions that use this have different ways to go about getting to the `ScopeRuleSets` in the first place (e.g. `matchShadowPseudoElementRules` looks at the `ShadowRoot::host` whereas `matchHostPseudoClassRules` looks at the `Element::shadowRoot` whereas etc.).  some of them even deal with multiple `RuleSet` (e.g. `matchSlottedPseudoElementRules`).

they're all also called from both `collectMatchingAuthorRules` and `matchUserRules` so we can't just hardcode `ScopeRuleSets::authorStyle` vs `ScopeRuleSets::userStyle` (in fact that's the very problem that this bug is trying to fix).  we could use an `enum class RuleSetOrigin` or something, but this is cleaner IMO

open to alternative suggestions if you have any :)

Reply:
I think you can just pass CascadeLevel (which should be renamed to CascadeOrigin) enum value and add a helper (ScopeRuleSets::ruleSetForCascadeOrigin or something) that returns the right RuleSet. 

This could also be a template argument though I doubt it matters.